src/Chapter3.hs

```diff
@@ -1007,6 +1150,89 @@ contestants, and write a function that decides the outcome of a fight!
 -}
 
 
+
+data MonsterAction
+    = MonsterAttack
+    | MonsterRunAway
+    deriving (Show)
+
+data KnightAction
+    = KnightAttack
+    | CastDefenceSpell Int
+    | DrinkHealthPotion Int
+    deriving (Show)
+
+data FighterKnight = FighterKnight
+    { fighterKnightName    :: String
+    , fighterKnightHealth  :: Int
+    , fighterKnightAttack  :: Int
+    , fighterKnightDefence :: Int
+    , fighterKnightActions :: [KnightAction]
+    } deriving (Show)
+
+data FighterMonster = FighterMonster
+    { fighterMonsterName    :: String
+    , fighterMonsterHealth  :: Int
+    , fighterMonsterAttack  :: Int
+    , fighterMonsterActions :: [MonsterAction]
+    } deriving (Show)
+
+class Fighter a where
+    getName :: a -> String
+    getDamaged :: a -> Int -> a
+    performNextAction :: (Fighter b) => a -> b -> (a, b)
+    isDead :: a -> Bool
+
+instance Fighter FighterMonster where
+    getName :: FighterMonster -> String
+    getName = fighterMonsterName
+
+    getDamaged :: FighterMonster -> Int -> FighterMonster
+    getDamaged m opponentAttack = m { fighterMonsterHealth = fighterMonsterHealth m - opponentAttack }
+
+    performNextAction :: (Fighter b) => FighterMonster -> b -> (FighterMonster, b)
+    performNextAction m opponent = case fighterMonsterActions m of
+        a:as  -> case a of
+            MonsterAttack -> (m { fighterMonsterActions = as ++ [a] }, getDamaged opponent (fighterMonsterAttack m))
```

Comment:
If you remember from Chapter Two, adding elements to the end of the list is not efficient. You can avoid this by calling the `cycle` function once at the beginning to get an infinite list of actions. And since Haskell is lazy, this won't be a problem :slightly_smiling_face:

Reply:
Oh yes! Thanks.